### DISCUSSION TITLE
Parallel execution dataloaders

### DISCUSSION THREAD

**Discussion by Dgsuser** [Netflix/dgs-framework#872] [Q&A]
Hi, I have a query which is like below..

query returning list of below details{
     MovieID : ID
     Director: director (implemented through dataloader)
     Reviews: reviews (implemented through dataloader)
}

director and reviews are two separate call not dependent on each other but by default this is executed sequentially one by one instead of parallel. Am using completable future and @dgsdata annotations for dataloader. Using spring webflux as well.

If we have multiple dataloaders how can we make it run parallel?

**Comment by srinivasankavitha** [comment]
The execution order is controlled by `graphql-java` so as far as I am aware
of,  there isn't a way to control the ordering for fields if you want them
in parallel.

**Comment by paulbakker** [comment]
@Dgsuser you need to make sure to pass in an executor when creating the completablefuture. If you do that, it will run on that executor (typically a worker thread pool), and things will be able to execute in parallel.

For example:

```
@DgsDataLoader(name = "messages")
public class MessageDataLoader implements BatchLoader<String, String> {

    private Executor executor = Executors.newCachedThreadPool();
    
    @Override
    public CompletionStage<List<String>> load(List<String> keys) {
        return CompletableFuture.supplyAsync(() -> keys.stream().map(key -> "hello, " + key + "!").collect(Collectors.toList()), executor);
    }
}

```

**Reply by jord1e** [reply]
> @Dgsuser you need to make sure to pass in an executor when creating the completablefuture. If you do that, it will run on that executor (typically a worker thread pool), and things will be able to execute in parallel.
> 
> For example:
> 
> ```
> @DgsDataLoader(name = "messages")
> public class MessageDataLoader implements BatchLoader<String, String> {
> 
>     private Executor executor = Executors.newCachedThreadPool();
>     
>     @Override
>     public CompletionStage<List<String>> load(List<String> keys) {
>         return CompletableFuture.supplyAsync(() -> keys.stream().map(key -> "hello, " + key + "!").collect(Collectors.toList()), executor);
>     }
> }
> ```

Doesn't `CompletableFuture#supplyAsync` already use a thread pool by default? 

https://github.com/openjdk/jdk/blob/4616c13c2f1ced8a8bdeed81f0469523932e91b5/src/java.base/share/classes/java/util/concurrent/CompletableFuture.java#L430-L438

https://github.com/openjdk/jdk/blob/4616c13c2f1ced8a8bdeed81f0469523932e91b5/src/java.base/share/classes/java/util/concurrent/CompletableFuture.java#L1988-L2006

> but by default this is executed sequentially one by one instead of parallel.
Maybe this is just a coincidence?

**Comment by paulbakker** [comment]
Added an issue to improve the docs: https://github.com/Netflix/dgs-framework/issues/873

**Comment by Dgsuser** [comment]
Hi Paul, Thank you for your reply. First, I would like to thank you for all the content created. The videos were really helpful. 🙏 

On this yes I agree if we use executor then it will be executed parallel. But when I posted this I was looking for parallel triggers from datafetcher class for all the dataloader existing like query execution. However this works as well I believe.

Thanks!
